QEfficient/generation/text_generation_inference.py

```diff
@@ -157,24 +158,26 @@ def cloud_ai_100_exec_kv_helper(
     # Skip inputs/outputs
     session.skip_buffers([x for x in session.input_names + session.output_names if x.startswith("past_")])
 
-    # Read prompt and ctx len from session
-    batch_size, _, ctx_len, _ = session.bindings[session.binding_index_map["past_key.0"]].dims
-    prefill_seq_len = max(
-        [x[session.binding_index_map["input_ids"]][1][1] for x in session.allowed_shapes]
-        + [session.bindings[session.binding_index_map["input_ids"]].dims[1]]
-    )
+    # Read batch_size and prefill_seq_len from session
+    if session.allowed_shapes:
+        batch_size = max([x[session.binding_index_map["input_ids"]][1][0] for x in session.allowed_shapes])
+        prefill_seq_len = max([x[session.binding_index_map["input_ids"]][1][1] for x in session.allowed_shapes])
+    else:
+        batch_size, prefill_seq_len = session.bindings[session.binding_index_map["input_ids"]].dims
 
     if len(prompt) < batch_size:
         prompt = prompt * -(batch_size // -len(prompt))  # Repeat prompt to required size
     prompt = prompt[:batch_size]  # Truncate prompts to required size
 
     inputs = tokenizer(prompt, return_tensors="np", padding=True)
     position_ids_update = inputs["attention_mask"].sum(1, keepdims=True)
-    if generation_len is None:
-        generation_len = ctx_len
     padded_len = inputs["input_ids"].shape[1]
     num_chunks = -(padded_len // -prefill_seq_len)  # ceil divide without float
     padded_len = num_chunks * prefill_seq_len  # Convert to a multiple of prompt_len
+    if generation_len is None:
+        if ctx_len is None:
+            raise ValueError("At least one of ctx_len or generation_len is needed")
+        generation_len = ctx_len - position_ids_update.max()
```

Comment:
Shouldn't we also check when `generation_len` is provided by user,
`asser generation_len> ctx_len-position_ids_update.max()` ?

Reply:
That case is captured in the if condition lineno:177 as above. We compute generation length only if user has not provided it.

Reply:
If there is space for 200 tokens, but the user wants to generate only 20 tokens, then the above assertion doesn't make sense. We only assert that the user wants to generate at least 1 token, which is what we're already checking, in the next line (181).

Reply:
As discussed, let's put a warning for this case.
And should we expose this `generation_len` variable for `infer`, `execute` APIs?

Reply:
I've added a warning for the generation_len. Also exposed `--generation_len`, in infer and execute APIs.